### PR TITLE
Remove the global `DeclarationEngine`

### LINKED ISSUE #2063: Remove the global `TypeEngine` and `DeclarationEngine` in favour of moving them into the `TypeCheckContext`
As mentioned [here](https://github.com/FuelLabs/sway/pull/2004#pullrequestreview-1012828683), this is an important follow-up to #2004.

It should make it much easier to enable type-checking the same program multiple times within the same process. Currently this might be possible, but doing so will endlessly bloat the global arena as it's never cleaned up.

Accumulating the type state from the beginning of `TypedProgram::type_check` and returning it within the `TypedProgram` will allow us to take advantage of Rust's ownership semantics and leave its cleanup to RAII.

### PR DESCRIPTION
**Pull request by emilyaherbert**
This PR removes the global `DeclarationEngine` in favor of keeping a local copy in the `TypeCheckContext`.

This PR is pretty big... there aren't any additional things added other than what I """ had to add """ in order to complete this task. At a high level:

- This PR introduces the `Engines` struct, which has two fields `type_engine: &TypeEngine` and `declaration_engine: &DeclarationEngine`.
- For most/many of the traits introduced in #3353, this PR changes these traits to take a `Engines` instead of a `&TypeEngine`

Closes #2063

### CODE REVIEW

**Review by JoshuaBatty — CHANGES_REQUESTED** [review]
So excited to see this land, thanks heaps for putting this together. 

I think there may have been some issues with rebasing onto master, at least what I can see with files is sway-lsp. There was an API refactor that happened in between you starting this. The LSP files in this PR don't match up with the current file layout / API. Might be worth taking a look at rebasing those files again.

File: sway-core/src/control_flow_analysis/flow_graph/mod.rs [hunk 1]
```diff
@@ -210,14 +226,109 @@ impl ControlFlowGraph {
...
+    ) -> ControlFlowGraphNode<'cfg> {
         ControlFlowGraphNode::EnumVariant {
-            variant_name: other.name.clone(),
+            variant_name: other_name,
             is_public,
         }
     }
 }
+
+// impl<'a, N, E, Ty, Ix> petgraph::visit::IntoEdgeReferences
```
[the remark this answers is not in the document]
Comment (emilyaherbert) [reply sway-core/src/control_flow_analysis/flow_graph/mod.rs:241]: Ah yes, thanks for catching this with all of the green on this PR!

**Comment by emilyaherbert** [comment]
@JoshuaBatty I've added the changes you requested by dismissed you "request for changes" review as you're OOO. Feel free to ask for additional changes though if you see anything that I should address.

**Review by mohammadfawaz — APPROVED** [review]
Amazing work @emilyaherbert! Thank you so much for this. It will be make a lot of people happy 😄

**Review by otrho — APPROVED** [review]

**Pull request merged**